### PR TITLE
[Headers]: Add headers component

### PR DESCRIPTION
**Pull request by maya**
## Description

Adds headers component to the site. 

https://standards-staging-add-headers.apps.cloud.gov/headers/

## TODO:

**Dev**
- [x] Separate out JS into standards components
- [x] Add media query variable: `$nav-width`
- [x] Figure out why mega menu is creating scrollbar on the right and left does not 
 - **fixed**: added `overflow-x: hidden` to `body` in 5a7fc58.
- [x] Change menu button, close button, search toggle from an anchor link to a button element (will have to remove button styles)
- [x] Add focus state to first item in the dropdown (close button) using JS
- [x] Add basic header with mega menu
- [x] Add `content: '|'` after list item, not after the anchor link
- [x] Add code samples
- [x] Add nav menu (simple and mega menu) as an include
- [x] Review class names to make sure they make sense and follow modified BEM @donjo 
- [ ] Remove jQuery @yozlet @rogeruiz 

**Design**
- [ ] Add documentation @jameshupp 
- [x] Add new screenshots @ericadeahl 

Resolves: #1312.

Before you hit Submit, make sure you’ve done whichever of these applies to you:

- [x] Follow the [18F Front End Coding Style Guide](https://pages.18f.gov/frontend/) and [Accessibility Guide](https://pages.18f.gov/accessibility/checklist/).
- [x] Run `npm test` and make sure the tests for the files you have changed have passed.
- [ ] Run your code through [HTML_CodeSniffer](http://squizlabs.github.io/HTML_CodeSniffer/) and make sure it’s error free.
- [x] Title your pull request using this format: [Website] - [UI component]: Brief statement describing what this pull request solves.

### CODE REVIEW

**Comment by maya** [comment]
Notes from @ericadeahl:

All headers:
- I think we should disable “department of web standards” link (it's disorienting to be taken to the WDS homepage)

Basic headers:
- move all the section links to the left 15 px so the blue drop down doesn’t touch the search bar when open

Extended headers:
- for the drop down, extend the blue box behind the section title vertically so it reaches the gray line above it

**Comment by ericadeahl** [comment]
A few comments from our chat today:

- [ ] Change "view" to "demo" on buttons

- [ ] Add a documentation section under each header (for component-specific documentation), as well as a text section after the intro text for any general header documentation 

- [ ] Add documentation and code for using an image logo as an alternative to using a text logo. The solution we discussed for this was adding a section in the documentation displaying the different logo options, then including the image logo as a commented out section in the code snippets for all four headers

- [ ] Add a fixed footer to the demo pages that links back to the WDS site

fixed footer


One more thing I noticed:

- [ ] When you have a drop down open, then you click on a section without a drop down, the open drop down does not close (but it should!) Is this related to the JS issue or a separate bug?

**Comment by maya** [comment]
> When you have a drop down open, then you click on a section without a drop down, the open drop down does not close (but it should!) Is this related to the JS issue or a separate bug?

@ericadeahl Yep, only clicking on the button closes it. We will need #1363 implemented in order to add this functionality. cc: @rogeruiz

**Comment by maya** [comment]
> Add documentation and code for using an image logo as an alternative to using a text logo. The solution we discussed for this was adding a section in the documentation displaying the different logo options, then including the image logo as a commented out section in the code snippets for all four headers

@ericadeahl I think we started talking about commenting it out in the HTML, but thought we ended just saying we would add info _just_ in the documentation. I may recommend this, just bc we are in a time crunch, and then we can do some testing around it later on. cc: @bradnunnally

**Comment by maya** [comment]
Our skipnav link points to `#main-content` but there's no `#main-content` on the page since it's not a page template. For example: https://standards-staging-add-headers.apps.cloud.gov/headers/basic/

Should we just add instructions to the documentation about this and leave the page as is?

cc: @nickbristow @msecret

**Comment by nickbristow** [comment]
@maya I would add a div with that id. Just so its clear.

**Comment by maya** [comment]
@nickbristow done in 06175e9.

File: config/gulp/javascript.js [hunk 1]
```diff
@@ -63,7 +63,7 @@ gulp.task(task, [ 'eslint' ], function (done) {
     .pipe(source('components.js'))
     .pipe(buffer())
     .pipe(sourcemaps.init({ loadMaps: true }))
-      .pipe(uglify())
+      // .pipe(uglify())
```
Comment (msecret) [inline config/gulp/javascript.js]: I'm guessing I shouldn't merge these changes right?

File: docs/_includes/navbar.html [hunk 2]
```diff
@@ -1,5 +1,5 @@
-<a class="skipnav" href="#main-content">Skip to main content</a>
-<header class="usa-site-header" role="banner">
+<a class="usa-skipnav" href="#main-content">Skip to main content</a>
+<header class="usa-header site-header" role="banner">
```
Comment (msecret) [inline docs/_includes/navbar.html]: Whats `site-header` for?

File: docs/doc_assets/css/styleguide.scss [hunk 3]
```diff
@@ -1506,3 +1408,40 @@ $font-light: 300;
     }
   }
 }
+
+// Demo fixed footer
+.demo-footer {
```
Comment (msecret) [inline docs/doc_assets/css/styleguide.scss]: Whats the demo footer? How does it relate to headers? Why doesn't it need all these separate CSS rules?

File: src/html/header-basic-mega.html [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    <div class="usa-disclaimer">
+      <div class="usa-disclaimer-inner">
+        <div class="usa-disclaimer-official">
+          <img src="{{ site.baseurl }}/assets/img/favicons/favicon-40.png" alt="U.S. flag">
+          <p>An official website of the United States government</p>
+        </div>
+      </div>
+    </div>
+    <div class="usa-nav-container">
+      <div class="usa-navbar">
```
Comment (msecret) [inline src/html/header-basic-mega.html]: What's the difference between `navbar` and `nav`? Is there a better name for navbar, I don't really understand it in this context.

File: src/html/header-nav-megamenu.html [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    </a>
+  </li>
+  <li>
+    <button class="usa-accordion-button usa-nav-link" aria-expanded="false" aria-controls="sidenav-3">
+      <span>Section title</span>
+    </button>
+    <ul class="usa-nav-submenu usa-megamenu" id="sidenav-3">
+      <div class="usa-grid-full">
+        <div class="usa-megamenu-col">
+          <li>
```
Comment (msecret) [inline src/html/header-nav-megamenu.html]: Is it OK semantically to have an `<li>` as not the child of a `<ul>`?

File: src/html/header-nav-megamenu.html [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    <a class="usa-nav-link" href="#">
+      <span>Section title</span>
+    </a>
+  </li>
+  <li>
+    <button class="usa-accordion-button usa-nav-link" aria-expanded="false" aria-controls="sidenav-3">
+      <span>Section title</span>
+    </button>
+    <ul class="usa-nav-submenu usa-megamenu" id="sidenav-3">
+      <div class="usa-grid-full">
```
Comment (msecret) [inline src/html/header-nav-megamenu.html]: Could the usa-grid-full be mixed in to `usa-megamenu-col` to reduce one extra div of structure?

File: src/js/components/header/search.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
+var $ = require('jquery');
```
Comment (msecret) [inline src/js/components/header/search.js]: Is it OK that this is jquery code? Whats the plan to merge it in with the jquery removal?

File: src/stylesheets/components/_accordions.scss [hunk 8]
```diff
@@ -81,6 +82,31 @@ $accordion-border: 3px solid $color-gray-lightest;
     > li {
       @include accordion-list-item-styles;
     }
+  }
+
+  [aria-expanded=false] {
```
Comment (msecret) [inline src/stylesheets/components/_accordions.scss]: These additions might be a merge conflict problem as this code should be in the mixins in the above file.

Did you want to make any changes to the accordion style? You might want to get rid of these changes and make your changes manually if it's hard to figure out the conflicts. I may have gotten something wrong when I was working on it with you.

File: src/stylesheets/components/_disclaimer.scss [hunk 9]
```diff
@@ -8,6 +8,16 @@
   }
 }
 
+.usa-disclaimer-inner {
```
Comment (msecret) [inline src/stylesheets/components/_disclaimer.scss]: What's the need for this new addition?

File: src/stylesheets/components/_header.scss [hunk 10]
```diff
@@ -0,0 +1,602 @@
+// Navigation ------------- //
+
+$height-nav-secondary: 5rem;
+$width-nav-sidebar:    250px;
+
+
+.usa-header {
+  width: 100%;
+  z-index: 1;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Are we setting z-indexes in variables yet? How are we keeping track that 1 makes sense here?

File: src/stylesheets/components/_header.scss [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+
+  @include media($nav-width) {
+    border-bottom: 1px solid $color-gray-light;
+  }
+
+  a {
+    border-bottom: none;
+  }
+
+  em {
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I don't quite understand what content `em` would be representing here. Why does a tag for emphasis have so many additoinal rules in the header?

File: src/stylesheets/components/_header.scss [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+
+    @include media($nav-width) {
+      float: right;
+      margin-bottom: 0;
+      max-width: 21.5rem;
+    }
+  }
+}
+
+.usa-navbar {
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: What do you think about putting all the nav related code in a separate `components/nav.scss` file?

File: src/stylesheets/components/_header.scss [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+    margin-left: 0;
+    margin-top: 3.3rem;
+  }
+
+  a {
+    color: $color-base;
+    text-decoration: none;
+  }
+
+  br {
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Are we using `<br>` tags for spacing? Or does this somehow cancel out moving something to the next line?

File: src/stylesheets/components/_header.scss [hunk 14]
```diff
@@ -0,0 +1,602 @@
...
+    display: none;
+
+    @include media($nav-width) {
+      display: block;
+    }
+  }
+}
+
+// scss-lint:disable PropertyCount
+.usa-menu-btn {
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Is there a reason this shouldn't go in `buttons.scss`?

File: src/stylesheets/components/_header.scss [hunk 15]
```diff
@@ -0,0 +1,602 @@
...
+  @include button-unstyled;
+  background-color: $color-primary;
+  color: $color-white;
+  display: inline;
+  float: right;
+  font-size: $h6-font-size;
+  line-height: 4rem;
+  height: 4rem;
+  width: 5.8rem;
+  text-align: center;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: These have to be in alphabetical order.

File: src/stylesheets/components/_header.scss [hunk 16]
```diff
@@ -0,0 +1,602 @@
...
+}
+// scss-lint:enable PropertyCount
+
+.usa-overlay {
+  @include position(fixed, 0);
+  background: $color-black;
+  opacity: 0;
+  transition: all 0.2s ease-in-out;
+  visibility: hidden;
+  z-index: 9999;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Same comment about z-indexes, how are we tracking them? This seems like it's a magic number.

File: src/stylesheets/components/_header.scss [hunk 17]
```diff
@@ -0,0 +1,602 @@
...
+  }
+}
+
+@mixin nav-border {
+  border-bottom: 0.7rem solid $color-primary;
+  padding-bottom: 1rem;
+}
+
+.usa-nav-link {
+  -moz-osx-font-smoothing: grayscale;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Could you explain the reason for these rules on these specific nav links? A comment might be good here.

File: src/stylesheets/components/_header.scss [hunk 18]
```diff
@@ -0,0 +1,602 @@
...
+  @include media($nav-width) {
+    @include outer-container();
+    @include padding(null $site-margins);
+    max-width: $site-max-width;
+  }
+}
+
+.usa-nav {
+  $sliding-panel-width: 26rem;
+  @include padding(2rem);
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I'm a little hesitant to use includes like this. What benefit does the include provide that pure css doesn't have?

File: src/stylesheets/components/_header.scss [hunk 19]
```diff
@@ -0,0 +1,602 @@
...
+  @include position(fixed, 0 0 0 auto);
+  @include size($sliding-panel-width 100%);
+  @include transform(translateX($sliding-panel-width));
+  background: $color-white;
+  border-left: 1px solid $color-gray-light;
+  border-right: 0;
+  display: flex;
+  flex-direction: column;
+  overflow-y: auto;
+  -webkit-overflow-scrolling: touch;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Does this rule have other browser equivalents?  What does it do?

File: src/stylesheets/components/_header.scss [hunk 20]
```diff
@@ -0,0 +1,602 @@
...
+        }
+      }
+    }
+  }
+
+  button {
+    @include button-unstyled;
+    font-weight: $font-normal;
+    padding: 1.4rem 1.5rem 1rem 1.8rem;
+    -moz-osx-font-smoothing: grayscale;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Why does just this button need to have these browser specific rules?

File: src/stylesheets/components/_header.scss [hunk 21]
```diff
@@ -0,0 +1,602 @@
...
+    &:hover {
+      background-color: $color-gray-lightest;
+      color: $color-primary;
+
+      @include media($nav-width) {
+        background-color: transparent;
+      }
+    }
+
+    &[aria-expanded=false] {
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Man of these rules seem like repeats from accordion. Do you think we should try and put some of it in a shared place to avoid repetition? Or is the code not similar enough?

File: src/stylesheets/components/_header.scss [hunk 22]
```diff
@@ -0,0 +1,602 @@
...
+
+    &:not(:last-child)::after {
+      @include media($nav-width) {
+        content: '|';
+        padding-left: 1rem;
+      }
+    }
+  }
+
+  a,
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Should these rules definitely apply to all `a` tags within secondary links?

File: src/stylesheets/components/_header.scss [hunk 23]
```diff
@@ -0,0 +1,602 @@
...
+    font-size: $h5-font-size;
+    text-decoration: none;
+
+    &:hover {
+      color: $color-primary;
+    }
+  }
+
+  .usa-header-search-button {
+    @include button-unstyled;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Is this repeat of the search bar component? Can we share reused code?

File: config/gulp/javascript.js (hunk 1 above)
Comment (maya) [inline config/gulp/javascript.js]: @rogeruiz should we remove this?

File: docs/_includes/navbar.html (hunk 2 above)
Comment (maya) [inline docs/_includes/navbar.html]: That is for the styleguide site header.

File: docs/doc_assets/css/styleguide.scss (hunk 3 above)
Comment (maya) [inline docs/doc_assets/css/styleguide.scss]: It's where the go to back to the site link is located on the demo page.

screen shot 2016-08-12 at 10 09 04 am

File: src/js/components/header/search.js (hunk 7 above)
Comment (maya) [inline src/js/components/header/search.js]: I think it's OK, and that we could do the removal of jQuery in another PR since we are waiting for #1357. cc @rogeruiz @yozlet

File: src/stylesheets/components/_disclaimer.scss (hunk 9 above)
Comment (maya) [inline src/stylesheets/components/_disclaimer.scss]: This is changed to have a centered grid rather than the disclaimer links floated to left and right.

File: src/stylesheets/components/_header.scss (hunk 10 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: No, we are not. It's all manual so far.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: `em` is the HTML element for logos that are text, which is what they should semantically use.

File: src/stylesheets/components/_header.scss (hunk 10 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Do you think we should add a system now? I think I remember seeing a couple of z-index values in the header file.

We could also just do variables in this header file to start, which would potentially make it easier to move them to site wide in another refactor.

File: src/stylesheets/components/_header.scss (hunk 13 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Not for spacing but for breaks on the logo text: on desktop the logo text is on two lines and on mobile it's on one line.

File: src/stylesheets/components/_header.scss (hunk 10 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Some ideas on how to create variable names:
- Based on the actual components: https://css-tricks.com/handling-z-index/
- Based on where in the stacking order; background, foreground, above ground, etc.

File: src/stylesheets/components/_header.scss (hunk 14 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: It's not related to all buttons but a button on the mobile header component so I don't think so. `buttons.scss` is used for the button component https://standards.usa.gov/buttons/.

File: src/stylesheets/components/_header.scss (hunk 16 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Same as above. We don't have a z-index system in place. Ideally, we should do some research on different z-index management systems and then choose one: https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=css%20z-index%20system

File: src/stylesheets/components/_header.scss (hunk 17 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: The only time you want to use antialiasing is light text on dark background which is what this is here. Sure we can add a comment.

File: src/stylesheets/components/_header.scss (hunk 20 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: It's the same reason above which is light text on dark background which comes into play on the active state of the button (white text + dark blue background).

File: src/stylesheets/components/_header.scss (hunk 22 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: I think it should: the container "secondary links" implies everything inside is a "link." But if you can think of a good reason not to and another approach, let me know.

File: src/stylesheets/components/_header.scss (hunk 23 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: It's not a repeat of anything else. It's the button that hide/shows the search form, which is only present in the header component.

File: src/stylesheets/components/_header.scss (hunk 21 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Here's the accordion code: https://github.com/18F/web-design-standards/blob/staging/src/stylesheets/components/_accordions.scss#L24

The only thing I see that's the same between them is `background-repeat: no-repeat;`

File: src/stylesheets/components/_header.scss (hunk 18 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: No reason in this case. Fixed in dc2235f.

File: src/stylesheets/components/_header.scss (hunk 15 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Fixed in ed49e5f a271ec8.

File: src/html/header-basic-mega.html (hunk 4 above)
Comment (maya) [inline src/html/header-basic-mega.html]: `nav` is all the navigation links. `navbar` is no nav links. It's the extended div the logo is in that makes up the navigation bar on mobile.

Here's some screenshots of it in mobile + desktop:

screen shot 2016-08-12 at 10 55 45 am

screen shot 2016-08-12 at 10 55 55 am

File: src/stylesheets/components/_accordions.scss (hunk 8 above)
Comment (maya) [inline src/stylesheets/components/_accordions.scss]: I'm wondering if this was just a rebase error that mashed the old accordion code with your refactor PR? I can't remember changing these.

File: src/stylesheets/components/_header.scss (hunk 19 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: I don't know what this does. It was code carried over from the site: https://github.com/18F/web-design-standards/blob/bd6f4a1c776eab2ff2565040a532423689fec3ac/docs/doc_assets/css/styleguide.scss#L267

Comment (maya) [inline src/stylesheets/components/_header.scss]: Ooh I found it here: https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-overflow-scrolling

>**touch**
Use momentum-based scrolling, where the content continues to scroll for a while after finishing the scroll gesture and removing your finger from the touchscreen. The speed and duration of the continued scrolling is proportional to how vigorous the scroll gesture was. Also creates a new stacking context.

https://css-tricks.com/snippets/css/momentum-scrolling-on-ios-overflow-elements/

Says it's non-standards. I think we should remove it.

Comment (maya) [inline src/stylesheets/components/_header.scss]: Removed in 8e56e89.

File: src/stylesheets/components/_header.scss (hunk 17 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Since it's used in multiple places in the repo, do you think we should still add a comment? Sounds more like a repeated pattern.

File: src/stylesheets/components/_header.scss (hunk 10 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Added the simple approach here: 8a92a09.

File: src/html/header-nav-megamenu.html (hunk 5 above)
Comment (maya) [inline src/html/header-nav-megamenu.html]: Nope, fixed in 34790e2.

File: src/stylesheets/components/_accordions.scss (hunk 8 above)
Comment (maya) [inline src/stylesheets/components/_accordions.scss]: Removed in 8fdc4fc.

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: There's a bit too much interaction between the header code and nav for it to make sense to put it in a separate file.

File: src/stylesheets/components/_header.scss (hunk 17 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I think we should because as a WDS developer, I still didn't know about it.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I find that confusing. As a developer, I don't think I'd understand that the `em` is supposed to be a logo.

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I think it needs to be split up in some way. This file is pretty massive right now, and I'm finding it pretty hard to understand. I think some sort of separation could help that.

File: src/stylesheets/components/_header.scss (hunk 17 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: Added in 1503b53.

Comment (msecret) [inline src/stylesheets/components/_header.scss]: Another option if it's a repeated matter: make a mixin for it that's well named and use that.

Comment (maya) [inline src/stylesheets/components/_header.scss]: Not sure if it's worth it bc I assume we're going to add postCSS for vendor prefixing, so we'd just have to change it and now we'd be making a mixin that we know we'll be deprecating.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: I'm going to add more info in our documentation: https://docs.google.com/document/d/1zMNdp5UqJlMUWmDOboirop2kBMg5U50cOfSOajA_Vt8/edit?disco=ou2agy5g7l5u

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: The nav and header are pretty entwined so not sure if we can accomplish this by separating files. I added additional comments to break it up and label the parts in: e6e4222. Does that help make it clearer?

File: src/stylesheets/components/_header.scss [hunk 24]
```diff
@@ -0,0 +1,631 @@
...
+
+.usa-nav {
+  $sliding-panel-width: 26rem;
+  @include position(fixed, 0 0 0 auto);
+  @include size($sliding-panel-width 100%);
+  @include transform(translateX($sliding-panel-width));
+  background: $color-white;
+  border-left: 1px solid $color-gray-light;
+  border-right: 0;
+  display: flex;
```
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Have we tested this in IE yet? Flexbox can be somewhat inconsistent in IE for version 10, 11.

Comment (maya) [inline src/stylesheets/components/_header.scss]: If flexbox doesn't work, it reverts to non-flexbox view which is just a different order of the search box to nav links on mobile. Since it functions, I think this is an OK degradation that's in line with progressive enhancement.

File: config/gulp/javascript.js (hunk 1 above)
Comment (rogeruiz) [inline config/gulp/javascript.js]: We should revert these changes so they don't get committed.

Comment (maya) [inline config/gulp/javascript.js]: Done in 70cdf7f.

File: src/js/components/header/search.js (hunk 7 above)
Comment (juliaelman) [inline src/js/components/header/search.js]: Adding note that #1357 has been merged into staging.

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (juliaelman) [inline src/stylesheets/components/_header.scss]: I think what Marco is saying is that we might want to start thinking about separating the header specific styles from the nav styles. This might help in creating the ability to using each of these sets of styles are their own. For example, we could possibly run into issues when want to use nav specific styles in another component, but are required to either remove or incorporate header styles.

File: circle.yml [hunk 25]
```diff
@@ -30,6 +30,11 @@ test:
     - ls -agolf _site/ # Ensure that build:website worked
 
 deployment:
+  add-headers:
```
Comment (juliaelman) [inline circle.yml]: Let's be sure to remove this new env and deployment prior to merging in. Happy to leave in for now, but not something we want to keep for long term.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (juliaelman) [inline src/stylesheets/components/_header.scss]: Agreed. Please do not use `<em>` for adding styling, as it is a styling element within itself and could possibly be confusing even with documentation.

File: docs/_includes/sidenav/mobile.html [hunk 26]
```diff
@@ -1,14 +1,14 @@
 {% assign current = page.url | downcase | split: '/' %}
 
-<aside class="sidenav-mobile">
-  <a class="sliding-panel-close" href="javascript:void(0);">
-    <img src="/assets/img/close.svg" alt="close">
-  </a>
-  <nav>
+<nav role="navigation" class="usa-nav sidenav-mobile">
```
Comment (juliaelman) [inline docs/_includes/sidenav/mobile.html]: Do these changes relate to the header or a solve for another issue? No big deal, but want to make sure to account for any changes that help solve any open issues we need to account for.

**Comment by juliaelman** [comment]
@maya nice work! A few notes here and some in the code:

1) The code samples are current not displaying correctly for all of the header examples.
2) As discussed earlier today, let's go ahead and add the header content in another PR that branches off of `add-headers` and submit it back against it. I'd rather not merge this in without the content, as that could possibly slow down a release if not merged in in a timely manner.
3) Tested in IE and came across some breakage on the basic headers and basic header mega menu in IE 9 [1]. The extended header styles are broken in all versions of IE [2]. 

There are also now some merge conflicts due to the jQuery removal and more than happy to help resolve those for you.

[1]
screen shot 2016-08-16 at 11 08 20 pm

[2] 
screen shot 2016-08-16 at 11 11 28 pm

**Comment by maya** [comment]
>1) The code samples are current not displaying correctly for all of the header examples.
Right on, I will open a separate PR for that.

>2) As discussed earlier today, let's go ahead and add the header content in another PR that branches off of add-headers and submit it back against it. I'd rather not merge this in without the content, as that could possibly slow down a release if not merged in in a timely manner.
👍 

>3) Tested in IE and came across some breakage on the basic headers and basic header mega menu in IE 9 [1]. The extended header styles are broken in all versions of IE [2].

I'm not seeing what you are in IE 9 [1]. Here's what I see:

screen shot 2016-08-16 at 8 45 28 pm

IE [2] Fixed in dd60540.

>There are also now some merge conflicts due to the jQuery removal and more than happy to help resolve those for you.

👍  @rogeruiz volunteered to help with these.

File: docs/_includes/sidenav/mobile.html (hunk 26 above)
Comment (maya) [inline docs/_includes/sidenav/mobile.html]: It's to use the same code as our new headers.

**Comment by maya** [comment]
@juliaelman merge conflict is resolved.

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: **tl;dr** there’s a lot going on with the headers. I recommend that we revisit this idea with validated user research. Implementing this is not something we shouldn’t do (I even have a couple of ideas), but only that it’s not as easy as it seems right now.

Now for the longer bit...

When we talk about the nav, it’s important to understand the foundation of the component and code. All navs start out the same: they’re essentially constructed as a mobile slide out nav and everything within it. This mobile nav then gets transformed into a horizontal oriented nav at large screen sizes, either the inset basic header or the extended header dropped below. It’s a single unit that functions together.

Although the markup is distinct, the functionality (i.e. responsive interaction via Sass) is interwoven. All these distinct pieces are orchestrated together in the stylesheets. The responsive behavior happens in the same code block, selector by selector.

See lines [6-9](https://github.com/18F/web-design-standards/blob/1a552233e5b017cb3a803217a34972ac89671bb9/src/stylesheets/components/_header.scss#L6...#L9), you’ll see three z-index variables for the header. Keeping them together ensures this is easy to manage and people know what is doing what and can find it in the same place. Splitting them apart would be confusing to manage. We wouldn’t put them in variables file bc they’re not global variables but scoped to the header.

>This might help in creating the ability to using each of these sets of styles are their own. For example, we could possibly run into issues when want to use nav specific styles in another component, but are required to either remove or incorporate header styles.

Before drawing to conclusions, I’d want to ask @ericadeahl. Has our user research validated this need? Do we have evidence of people wanting to use the navigation in another component that's not the header?

Using a header with a nav, and then a nav again somewhere else on the page will cause conflict bc the code is a mobile nav and that would overlap with each other.

On desktop, We use a custom variable `$nav-width`, if you’re using the header and then you want to use the nav for another use case, you can’t change this variable to something else. They will be the same, but you don’t want them to be the same, you’d want to be able to control the width for your use case. The current code doesn’t support that. You’d have to re-copy this code and then add a new variable, you can see this variable is used many times throughout the code.

@juliaelman @msecret

**Comment by maya** [comment]
As of today 8/18/16, all the existing issues have been tackled listed in this PR and it has been fully reviewed. The remaining work will be coming via new PR's to the `add-headers` branch.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: Was this addressed @maya ?

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (msecret) [inline src/stylesheets/components/_header.scss]: I think we should talk about this in person/virtually as there's a lot of discussion going on here. I generally think this is not just about user need of using the nav but about code organization and modularization.

File: src/stylesheets/components/_header.scss (hunk 11 above)
Comment (maya) [inline src/stylesheets/components/_header.scss]: @msecret not yet. 

>as it is a styling element within itself

What does this mean? `em` is a semantic element which has emphasis?

Comment (maya) [inline src/stylesheets/components/_header.scss]: Adding a class in: f9bb134.

Although by applying this to the semantically correct element which we want people to use, the `em`, we don't have to enforce this.

File: src/stylesheets/components/_header.scss (hunk 12 above)
Comment (juliaelman) [inline src/stylesheets/components/_header.scss]: @maya @msecret and myself met this afternoon to discuss a path moving forward. We are in agreement in leaving this as is for now, but that there eventually will be `components/nav.scss` file.
